### DISCUSSION TITLE
Workflows for rebasing all changes after a fetch?

### DISCUSSION THREAD

**Discussion by bcspragu** [jj-vcs/jj#5216] [General]
**TL;DR - What's the equivalent of [git-branchless](https://github.com/arxanas/git-branchless/)'s `git sync` in `jj`?**

Longer explanation:

I frequently have five or six anonymous branches that I'm working on in a given repo, e.g.

```
@  ooozwmnq brandon@sprague.mx 2024-12-30 20:47:29 c0505549
│  Change #5
│ ○  vywnqtkx brandon@sprague.mx 2024-12-30 20:47:26 df4a26df
├─╯  Change #4
│ ○  vslmknsz brandon@sprague.mx 2024-12-30 20:47:15 bc0eeff7
├─╯  Change #3
│ ○  lqwuuksv brandon@sprague.mx 2024-12-30 20:47:13 98fe18c2
├─╯  Change #2
│ ○  lyoszovo brandon@sprague.mx 2024-12-30 20:46:58 8a60a6c1
├─╯  Change #1
○  mkqwwrry brandon@sprague.mx 2024-12-30 20:46:43 main b1750983
│  Create README.md
◆  zzzzzzzz root() 00000000
```

Now I merge in `Change #1` (or I just fetch other people's changes), and `main` moves, e.g.

```
@  ssopxytu brandon@sprague.mx 2024-12-30 20:53:24 d82fb96d
│  (empty) (no description set)
◆  lyoszovo brandon@sprague.mx 2024-12-30 20:46:58 main 8a60a6c1
│  Change #1
│ ○  ooozwmnq brandon@sprague.mx 2024-12-30 20:47:29 c0505549
├─╯  Change #5
│ ○  vywnqtkx brandon@sprague.mx 2024-12-30 20:47:26 df4a26df
├─╯  Change #4
│ ○  vslmknsz brandon@sprague.mx 2024-12-30 20:47:15 bc0eeff7
├─╯  Change #3
│ ○  lqwuuksv brandon@sprague.mx 2024-12-30 20:47:13 98fe18c2
├─╯  Change #2
◆  mkqwwrry brandon@sprague.mx 2024-12-30 20:46:43 b1750983
│  Create README.md
~
```

What workflow are people using to rebase all those changes back onto `main`? For example:

1. Do people just manually `jj rebase -s <first revision> -d main` for each branch?
2. Something like `jj rebase -b 'all:heads(all() ~ main)' -d main` gets the job done (no idea if that's the best way to do it, I'm bad at revsets)
3. On one project, I have some alias like:

```bash
jj rebase -s 'all:roots(main..bookmarks(glob:"brandon/*"))' -d main --ignore-immutable
```

And then there's the approach [Chris Krycho uses](https://v5.chriskrycho.com/notes/jj-tip-rebase-all-branches-matching-a-revset/), where he has a `wip` revision with all his actual branches as parents, and then he does `jj rebase -b all:wip -d main`

All of these work, but I'm wondering if there's some canonical way that makes sense for the common case and people recommend for one reason or another.

**Comment by PhilipMetzger** [comment]
There's #1039 for that command and @essiene is working on it. It also is a long desired feature for a longtime user like me. It does mean that you'll need to use the workarounds you posted above until exists.

**Reply by bcspragu** [reply]
Ah excellent, thanks for the info!

**Comment by jrmoulton** [comment]
This is what I use. Seems to work pretty well
```
jj git fetch --all-remotes
jj rebase -s 'all:roots(mutable())' -d 'trunk()'
```
